plugins/crd/crd_queue.go

```diff
@@ -49,39 +53,53 @@ func networkserviceWork(plugin *Plugin) {
 		// We define a function here to process a queue item, so that we can
 		// use 'defer' to make sure the message is marked as Done on the queue
 		func(key string) {
-			var obj interface{}
-			var err error
-
 			defer queueNS.Done(key)
 
 			// Attempt to split the 'key' into namespace and object name
 			namespace, name, err := cache.SplitMetaNamespaceKey(strKey)
 
 			if err != nil {
+				// This is a soft-error, retry up to QUEUE_RETRY_COUNT times
 				plugin.Log.Errorf("Error splitting meta namespace key into parts: %s", err.Error())
-				runtime.HandleError(fmt.Errorf("Error splitting meta namespace key into parts: %s", err.Error()))
-				// This is a soft-error, we merely want to forget it and mark it done on the queue
-				queueNS.Forget(key)
+				if err != nil {
+					if queueNS.NumRequeues(key) < QUEUE_RETRY_COUNT {
+						plugin.Log.Errorf("Error processing item with key %s, error %v", key, err)
+						queueNS.AddRateLimited(key)
+					} else {
+						plugin.Log.Errorf("Failed processing item with key %s, error %v, no more retries", key, err)
+						queueNS.Forget(key)
+					}
+				}
 				return
 			}
 
 			plugin.Log.Infof("Read item '%s/%s' off workqueue. Processing...", namespace, name)
 
-			// Retrieve the latest version in the cache of this alert
-			obj, err = plugin.sharedFactoryNS.Networkservice().V1().NetworkServices().Lister().NetworkServices(namespace).Get(name)
+			// Retrieve the latest version in the cache of this NetworkService. By using
+			// GetByKey() we are able to determine if the item exists, and thus if it was
+			// added or deleted from the queue, and process appropriately
+			item, exists, err := plugin.informerNS.GetIndexer().GetByKey(strKey)
```

Comment:
What kind of cache is this? Is it possible for the key to be evicted while the object may actually exist?

Reply:
These are SharedIndexInformer caches, see here [1] and here [2]. It should not be possible for the key to evicted while the object actually exists either, unless there is an error in the processing logic.

[1] https://github.com/ligato/networkservicemesh/blob/master/plugins/crd/plugin_impl_crd.go#L68
[2] https://github.com/ligato/networkservicemesh/blob/master/plugins/crd/plugin_impl_crd.go#L251